### DISCUSSION TITLE
`pnpm dlx` behaving slightly differently than `npx`

### DISCUSSION THREAD

**Discussion by virtuoushub** [pnpm/pnpm#5820] [Q&A]
# Question

Is the expected behavior `pnpm dlx` errors and `pnpm exec` works when attempting dynamic package execution? 

The [feature comparison](https://pnpm.io/feature-comparison) mentions `pnpm dlx` as the way to match `npx` for _dynamic package execution_ (thanks for point me to this doc @meduzen!).

_note_ using `pnpm exec` works for the below example. So `pnpm exec playwright install --with-deps` is behaving as I need it to; I am just curious if I missed something in the documentation.

_Small aside_: Is what I ran into something specific to how `playwright` bundles itself? That is to say `pnpm dlx` is behaving as designed, it's just this particular package (`playwright`) is the offending party?

---

## Context

I am currently seeing:

```
╔════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║ Playwright Test integrity check failed:                                                                ║
║ You have @playwright/test version '1.29.0' and 'playwright' version '1.29.1' installed!                ║
║ You probably added 'playwright' into your package.json by accident, remove it and re-run 'npm install' ║
╚════════════════════════════════════════════════════════════════════════════════════════════════════════╝
```

after some research, I came across:

> Seems like the issue is related to the use of `pnpm` and the `pnpx` command, after changing to use `npx` it no longer throws the error
> 
> ```
> pnpx playwright install --with-deps < throws error
> 
> npx playwright install --with-deps < works fine
> ```

See: https://github.com/microsoft/playwright/issues/17955#issuecomment-1277376213

The referenced GitHub issue was using `pnpx` but the same goes for `pnpm dlx playwright install --with-deps` (that throws an error)

---

See also: 

* https://pnpm.io/cli/dlx
* https://pnpm.io/cli/exec
* https://pnpm.io/feature-comparison

**Comment by zkochan** [comment]
Yes, looks like you need to run either:

```
pnpm exec playwright install --with-deps
```

or

```
pnpm playwright install --with-deps
```

**Reply by virtuoushub** [reply]
Thanks for the reply!

Is there some documentation I missed as to what the expected behavior of `pnpm dlx` and `pnpm exec` are?

I expected `pnpm dlx` to work _basically_ the same as `npx`, but that wasn't the case here.

**Reply by meduzen** [reply]
Same on my side, especially because the [feature comparison](https://pnpm.io/feature-comparison) mentions `pnpm dlx` as the way to match `npx` for _dynamic package execution_.

It seems indeed that `pnpm exec` is the way to go.

**Reply by simwai** [reply]
Very confusing that pnpx and pnpm exec behaviour difference
